libnavigation-core/src/test/java/com/mapbox/navigation/core/replay/history/ReplayHistoryPlayerTest.kt

```diff
@@ -0,0 +1,155 @@
+                    ReplayEventLocation(
+                        lat = 49.2450478,
+                        lon = 8.8682922,
+                        time = 1580777820.952,
+                        speed = 30.239412307739259,
+                        bearing = 108.00135040283203,
+                        altitude = 222.47210693359376,
+                        accuracyHorizontal = 3.9000000953674318,
+                        provider = "fused")),
+                ReplayEventUpdateLocation(1580777822.959,
+                    ReplayEventLocation(
+                        lat = 49.2448858,
+                        lon = 8.8690847,
+                        time = 1580777822.958,
+                        speed = 29.931121826171876,
+                        bearing = 106.001953125,
+                        altitude = 221.9241943359375,
+                        accuracyHorizontal = 3.9000000953674318,
+                        provider = "fused")),
+                ReplayEventUpdateLocation(1580777824.953,
+                    ReplayEventLocation(
+                        lat = 49.2447354,
+                        lon = 8.8698759,
+                        time = 1580777824.89,
+                        speed = 29.96711540222168,
+                        bearing = 106.00138092041016,
+                        altitude = 221.253662109375,
+                        accuracyHorizontal = 3.9000000953674318,
+                        provider = "fused"))
+            )
+        )
+        val replayHistoryPlayer = ReplayHistoryPlayer(replayHistoryData)
+        replayHistoryPlayer.observeReplayUpdates(mockLambda)
+
+        val job = replayHistoryPlayer.play(lifecycleOwner)
+        coAdvanceTimeSeconds(3)
+        replayHistoryPlayer.finish()
+        job.cancelAndJoin()
+
+        // Note that it only played 2 of the 3 locations
+        val replayUpdates = mutableListOf<ReplayUpdate>()
+        coVerify { mockLambda.invoke(capture(replayUpdates)) }
+        val events = replayUpdates.flatMap { it.replayEventHappened }
+        assertEquals(events.size, 2)
+        assertEquals(events[0].event_timestamp, 1580777820.952)
+        assertEquals(events[1].event_timestamp, 1580777822.959)
+    }
+
+    @Test(expected = Exception::class)
+    fun `should crash if history data is empty`() {
+        val replayHistoryData = ReplayHistoryData(
+            listOf())
+        val replayHistoryPlayer = ReplayHistoryPlayer(replayHistoryData)
+
+        replayHistoryPlayer.play(lifecycleOwner)
+        replayHistoryPlayer.finish()
+    }
+
+    /**
+     * Helpers for moving the simulation clock
+     */
+
+    @Before
+    fun setup() {
+        mockkStatic(SystemClock::class)
+        every { SystemClock.elapsedRealtimeNanos() } returns deviceElapsedTimeNanos
+    }
+
+    @After
+    fun teardown() {
+        unmockkObject(SystemClock.elapsedRealtimeNanos())
+    }
+
+    private suspend fun coAdvanceTimeSeconds(seconds: Int) {
```

Comment:
Why do we need to create our advance helper function instead of using [`advanceTimeBy`](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/kotlinx.coroutines.test/-delay-controller/advance-time-by.html)?

Reply:
I considered making a TestRule for this (ended with thinking it's over design at the moment). But we need to advance the SystemClock, the coroutine test rule, and delay the coroutine - all to simulate advancing time instead of actually advancing time

These ideas were built off the research for making [MapboxTimerTest](https://github.com/mapbox/mapbox-navigation-android/blob/master/libnavigation-util/src/test/java/com/mapbox/navigation/utils/timer/MapboxTimerTest.kt#L27)